Nebari-CLI project internship blog-post 2022

asmijafar20 opened the pull request:
<!--Hi there! If you aren't submitting a blog post, you can delete this message.
Thank you for submitting a blog post! We want to make sure content produced for Quansight Labs is accessible, so please address the the following guidelines in your post.-->

## Text styling

- [ ] The blog is written with [plain language](https://www.plainlanguage.gov/guidelines/) (where relevant).
- [ ] If there are headers, they use the proper header tags in order to do so (with only one level-one header).
- [ ] All links describe where they link to (for example, check the [Quansight labs website](https://labs.quansight.org/)).
- [ ] Any kind of styling that the author uses (for example, bold for emphasis) is consistent throughout the blog.

## Non-text contents

- [ ] All content is represented as text (for example, images need alt text and videos need captions or descriptive transcripts).
- [ ] If there are emojis, there are not more than three in a row.
- [ ] Don't use [flashing gifs or videos](https://www.w3.org/TR/UNDERSTANDING-WCAG20/seizure-does-not-violate.html).
- [ ] If it were to be read as plain text, the blog still makes sense and no information is missing.

melissawm commented:
@noatamir could you help us understand why this preview is failing? thanks!

noatamir commented:
I need to add all the authors in StoryBlock CMS in advance to generate their author slug. I just did this for @asmijafar20, and I am making it for the others as well. A blog post won't generate without having the author in the website CMS in advance.

asmijafar20 commented:
Hey @noatamir, It's failing again.

iameskild commented on apps/labs/posts/nebari-typer-cli.md:
```diff
@@ -0,0 +1,170 @@
+## ✅ Initial steps
+
+Nebari already has a command line interface with all the functionalities but the structure of the CLI was not good and it was quite messy. I have to work on making it more clearer, structured and pretty. So, I started all the tools and libraries that we can use to make a CLI better. After exploring various tools and libraries it was decided to work with [Typer](https://typer.tiangolo.com/) and [R ...
+Initially it was hard to implement `typer` on the predefined functions of the CLI as it was important to understand how all functions fits together and their functionalities.
```
```suggestion
Prior to implementing the new Typer-based CLI, I spent some time trying to understand what each of the commands do and where the changes needed to be made. Once this was understood, I created a minimum viable product (MVP), shared it with the team and got the green-light to proceed. 
```

iameskild commented on apps/labs/posts/nebari-typer-cli.md:
```diff
@@ -0,0 +1,170 @@
+
+Nebari already has a command line interface with all the functionalities but the structure of the CLI was not good and it was quite messy. I have to work on making it more clearer, structured and pretty. So, I started all the tools and libraries that we can use to make a CLI better. After exploring various tools and libraries it was decided to work with [Typer](https://typer.tiangolo.com/) and [R ...
+Initially it was hard to implement `typer` on the predefined functions of the CLI as it was important to understand how all functions fits together and their functionalities.
+In the next section, we will see how the new Nebari CLI looks like with all it's functionalities.
```
```suggestion
In the next section, we will take a look at this new Nebari CLI.
```

iameskild commented on apps/labs/posts/nebari-typer-cli.md:
```diff
@@ -0,0 +1,170 @@
+    <i>Output of nebari init help command</i>
+</p>
+
+Like we can see it has lots of options available to pass as a flag in the command while executing it. The `guided-init` flag in the init command is the most valuable flag. If we are using Nebari first time and we don't know exactly what other options and arguments are there for the `init` command then we can run `init` command using `--guided-init` flag. It's ask values from the user and it is mo ...
```
```suggestion
We can see that the `nebari init` command has a lot of available options. If we are using Nebari first time and we don't know exactly what other each of the options represent, then running `nebari init --guided-init` will guide you through a simple questionnaire which provides helpful information ,and links to additional resources, to assist the end-user with a number of important decisions.
```

iameskild commented on apps/labs/posts/nebari-typer-cli.md:
```diff
@@ -0,0 +1,170 @@
+    <i>Output of nebari init guided-init command</i>
+</p>
+
+As we can see it's more like a survey that asks users to fill out the correct details for generating the `nebari-config.yaml` file. We have used the [questionary library](https://github.com/tmbo/questionary) for having the options for the questions.
```
```suggestion
As we can see it's more like a survey that asks users to fill out the details for generating the `nebari-config.yaml` file. To create this questionnaire, we used the [questionary library](https://github.com/tmbo/questionary). 
```

iameskild commented on apps/labs/posts/nebari-typer-cli.md:
```diff
@@ -0,0 +1,170 @@
+
+As we can see it's more like a survey that asks users to fill out the correct details for generating the `nebari-config.yaml` file. We have used the [questionary library](https://github.com/tmbo/questionary) for having the options for the questions.
+
+After creating or generating the `nebari-config.yaml`, It's time to validate and render it respectively. First we will see the `help` command output for the validate to see all the options and requirements available for it.
```
```suggestion
After generating the `nebari-config.yaml`, it's time to validate and render it respectively. First we will see the `help` command output for the validate to see all the options and requirements available for it.
```

iameskild commented on apps/labs/posts/nebari-typer-cli.md:
```diff
@@ -0,0 +1,170 @@
+    <i>Output of nebari validate and render command</i>
+</p>
+
+Render command builds the infrasture for the deployment. It creates all terraform stages file that are required during deployment.
```
```suggestion
Render command renders all of the Terraform scripts used for the actual deployment. These Terraform scripts are broken into stages and are stored in directory called `stages`.
```

iameskild commented on apps/labs/posts/nebari-typer-cli.md:
```diff
@@ -0,0 +1,170 @@
+
+Render command builds the infrasture for the deployment. It creates all terraform stages file that are required during deployment.
+
+**_NOTE:_** It's not neccessary to run validate and render command seperately. Because it runs under the hood of the deploy command.
```
```suggestion
**_NOTE:_** It's not necessary to run validate and render command separately as both commands are run under the hood by the deploy command.
```

iameskild commented on apps/labs/posts/nebari-typer-cli.md:
```diff
@@ -0,0 +1,170 @@
+
+**_NOTE:_** It's not neccessary to run validate and render command seperately. Because it runs under the hood of the deploy command.
+
+Now we will see the ouput of the `nebari-deploy` help command. Deploy command creates the provisioned resources that Nebari requires and deploys your project. Nebari also has a destroy command that works the same way the deploy works but instead of creating the provisioned resources it destroys it. It comes with the confirmation that is very important feauture that the new CLI have because previo ...
```
```suggestion
Now we will see the output of the `nebari deploy` help command. The deploy command creates the provisioned resources that Nebari requires and deploys your project. Nebari also has a destroy command that works the same way the deploy works but instead of creating the provisioned resources it destroys it. The destroy command now requires confirmation, an important new feature. Previously it didn't ask for the confirmation and the cluster would start being destroyed right after executing it.
```

iameskild commented on apps/labs/posts/nebari-typer-cli.md:
```diff
@@ -0,0 +1,170 @@
+    <i>Output of nebari deploy help command and destroy command</i>
+</p>
+
+It has many flag which user can pass while executing the deploy command according to it's preferences. Here also `-c` flag is important to pass to make the deployment of `nebari-config.yaml` file successful.
```
```suggestion
The deploy command also has a `--config` flag.
```

I removed the sentence about additional flags here because one of the potential enhancements we want to make in the future is to remove `--dns-provider` and `--dns-auto-provision` and move them to the `init` stage. We just didn't have the time for it this time around.

iameskild commented on apps/labs/posts/nebari-typer-cli.md:
```diff
@@ -0,0 +1,170 @@
+## 😇 My learnings and Acknowledgement
+
+I learned a lot during my internship. First thing first I learned about `Typer` and `Rich` and their functionalities. Interacting with the command line interface is easy but understanding and making it more structured and user-friendly can be tough in the start but playing around with functions and it's features can make it easy.
+Nebari is very usefule tool for the data scientists/reseachers to develop their project. I hope to continue working on this very interesting project that makes life so much easier without having the much knowledge about devOps.
```
```suggestion
Nebari is very useful platform for data scientists/researchers to develop their project. I hope to continue working on this very interesting project that makes life so much easier without requiring in depth devOps knowledge.
```

iameskild reviewed (changes requested):
This is looking great @asmijafar20! Thank you for all your hard work. 

I had several suggestions below, mostly just making sure the sentences flowed a bit more smoothly. 

I also wanted to point out that all of the screenshots have `qhub` in them (perhaps that's the folder you're working in) but if you have time to take new ones (without any reference to `qhub`), that would be much appreciated :)

Thanks again @asmijafar20

trallard commented on apps/labs/posts/nebari-typer-cli.md:
```diff
@@ -0,0 +1,161 @@
+    <i>Output of nebari validate and render command</i>
+</p>
+
+Render command builds the infrasture for the deployment. It creates all terraform stages file that are required during deployment.
```
```suggestion
The render command builds the infrasture for the deployment. It creates all terraform stages file that are required during deployment.
```

+ all other suggestions

trallard commented on apps/labs/posts/nebari-typer-cli.md:
```diff
@@ -0,0 +1,161 @@
+
+Render command builds the infrasture for the deployment. It creates all terraform stages file that are required during deployment.
+
+**_NOTE:_** It's not neccessary to run validate and render command seperately. Because it runs under the hood of the deploy command.
```
```suggestion
**_NOTE:_** It's not necessary to run the `validate` and `render` commands separately. Because it runs under the hood of the deploy command.
```
`+` the other suggestions

trallard commented on apps/labs/posts/nebari-typer-cli.md:
```diff
@@ -0,0 +1,161 @@
+
+## 😇 My learnings and Acknowledgement
+
+I learned a lot during my internship. First thing first I learned about `Typer` and `Rich` and their functionalities. Interacting with the command line interface is easy but understanding and making it more structured and user-friendly can be tough in the start but playing around with functions and it's features can make it easy.
```
```suggestion
I learned a lot during my internship. First, I learned about `Typer` and `Rich` and its functionalities. Interacting with the command line interface was easy but understanding how it worked and making it more structured and user-friendly was tough initially, but playing around with the respective functions and features made it more accessible.
```

asmijafar20 commented:
@iameskild and @viniciusdc Please review the PR. Suggest the changes if you find any. Thanks. 
cc: @trallard

iameskild commented on apps/labs/posts/nebari-typer-cli.md:
```diff
@@ -0,0 +1,161 @@
+
+Below is the diagram of the Nebari CLI structure.
+
+<img alt="Nebari commands" src="/posts/asmi-blogposts/nebari-commands.png" />
```
@pavithraes and I iterated on this a bit [here](https://docs.google.com/presentation/d/1lSeydRDhhqZ9mhZHtOJlKSrmpIWCBaKyLU9GyRxq1So/edit#slide=id.g15b6aade2db_1_78). 

It's not perfect, but I think it highlights the most important commands and what the result of each command is.

noatamir replied:
<img width="1099" alt="Updated image" src="https://user-images.githubusercontent.com/6564007/199562342-0b3d66b8-e0ac-428e-bcf2-4afb7ac4cc2b.png">
Thanks for sharing the file @asmijafar20! I tweaked this image a bit because I noticed it was pixelated. @trallard would you like me to make additional changes? My main concerns were the nebari logo's proportions and crispness, but I retouched the arrow and cloud position while I was at it. I can undo it, or make other changes as you prefer.

@asmijafar20 If you would like to share the files you use to generate the other images/designs where the nebari logo appears (hero and feature images), I can repeat the same tweaks. If Tania confirms it, then I believe those are the last changes 😇

asmijafar20 replied:
I have uploaded the hero and feature images in the drive link of the template images.

noatamir replied:
It's not in that folder now 😿  can you share it with me again?

asmijafar20 replied:
I don't have access to the drive folder now. What to do now?

iameskild commented on apps/labs/posts/nebari-typer-cli.md:
```diff
@@ -0,0 +1,162 @@
+
+## 🪴 About Nebari
+
+Nebari is an open source project from Quansight that enables teams/users to deploy a cost-effective and scalable compute/data science platforms in the cloud. It uses an infrastructure-as-a-code approach to quickly and smoothly deploy a shared data science environment, on-premises or in the cloud of choice. For more details, refer to the [Nebari docs](https://www.nebari.dev/).
```
```suggestion
Nebari is an open source project from Quansight that enables users to deploy a cost-effective and scalable data science platforms in the cloud. It uses an infrastructure-as-a-code approach to quickly and smoothly deploy a shared data science environment, on-premises or in the cloud of choice. For more details, refer to the [Nebari docs](https://www.nebari.dev/).
```

iameskild commented on apps/labs/posts/nebari-typer-cli.md:
```diff
@@ -0,0 +1,162 @@
+
+## Nebari Command-line Interface (CLI)
+
+I've worked on developing the Nebari [Command-line interface (CLI)](https://en.wikipedia.org/wiki/Command-line_interface) during my three months internship.
```
```suggestion
I worked on developing the Nebari [Command-line interface (CLI)](https://en.wikipedia.org/wiki/Command-line_interface) during my three months internship.
```

iameskild commented on apps/labs/posts/nebari-typer-cli.md:
```diff
@@ -0,0 +1,162 @@
+
+I've worked on developing the Nebari [Command-line interface (CLI)](https://en.wikipedia.org/wiki/Command-line_interface) during my three months internship.
+
+The Nebari CLI consists of various commands the user needs to run to deploy, configure, and update Nebari. Let's see briefly what each one of the `nebari` commands does.
```
```suggestion
The Nebari CLI consists of various commands the user needs to run to initialize, deploy, configure, and update Nebari. Let's see briefly what each one of the `nebari` commands does.
```

iameskild commented on apps/labs/posts/nebari-typer-cli.md:
```diff
@@ -0,0 +1,162 @@
+
+## ✅ Initial steps
+
+Prior to my internship, Nebari already has a command line interface with all the functionality listed above however it was clunky and, at times, a little challenging to use. My task as an intern was to improve the CLI experience for the end-user. After exploring various tools and libraries it was decided that we would proceed with [Typer](https://typer.tiangolo.com/) and [Rich](https://rich.readt ...
```
```suggestion
Prior to my internship, Nebari had an [`argparse`-based](https://docs.python.org/3/library/argparse.html) command line interface which included all the functionality listed above. However, it was clunky and at times, a little challenging to use. My task as an intern was to improve the CLI experience for the end-user. After exploring various tools and libraries it was decided that we would proceed with [Typer](https://typer.tiangolo.com/) and [Rich](https://rich.readthedocs.io/en/stable/introduction.html). For those unfamiliar, Typer is a Python library that makes CLI applications more interactive and user friendly. I started exploring features of `typer` and how it could be implemented as a replacement for the current CLI.
```

iameskild commented on apps/labs/posts/nebari-typer-cli.md:
```diff
@@ -0,0 +1,162 @@
+    <i>Output of the nebari init help command</i>
+</p>
+
+We can see that the `nebari init` command has a lot of available options. If we are using Nebari first time and we don't know exactly what other each of the options represent, then running `nebari init --guided-init` will guide us through a simple questionnaire which provides helpful information ,and links to additional resources, to assist with a number of important decisions.
```
```suggestion
The `nebari init` command has many available options. If a user is trying Nebari for the first time, it's hard to know what other each of these options represent. As a result, we implemented a helpful guided initialization wizard: `nebari init --guided-init`. This command guides the user through a simple questionnaire to provide helpful information and links for a number of important decisions that the user needs to make.
```

iameskild commented on apps/labs/posts/nebari-typer-cli.md:
```diff
@@ -0,0 +1,162 @@
+    <i>Output of nebari init guided-init command</i>
+</p>
+
+As we can see it's more like a survey that asks end-users to fill out the details for generating the `nebari-config.yaml` file. To create this questionnaire, We have used the [questionary library](https://github.com/tmbo/questionary).
```
```suggestion
This guided initialization wizard (equivalent to using `nebari init` with the appropriate flags) generates the all important `nebari-config.yaml` file. To help us generate this questionnaire, we used a package called [questionary](https://github.com/tmbo/questionary).
```

iameskild commented on apps/labs/posts/nebari-typer-cli.md:
```diff
@@ -0,0 +1,162 @@
+
+As we can see it's more like a survey that asks end-users to fill out the details for generating the `nebari-config.yaml` file. To create this questionnaire, We have used the [questionary library](https://github.com/tmbo/questionary).
+
+After generating the `nebari-config.yaml`, it's time to validate and render it respectively. First we will see the `help` command output for the validate to see all the options and requirements available for it.
```
```suggestion
After generating the `nebari-config.yaml`, it's time to validate and render the Terraform scripts. To inspect how to use the validate and render commands, let's have a look at their respective `--help` commands.
```

iameskild commented on apps/labs/posts/nebari-typer-cli.md:
```diff
@@ -0,0 +1,162 @@
+    <i>Output of the nebari validate and render help command</i>
+</p>
+
+As we can see we have to pass the `-c/--config` flag for validating the `nebari-config.yaml` file and we have to do the same for rendering the infrastructure using the render command.
```
```suggestion
We need to pass the `-c/--config` flag to validate the `nebari-config.yaml` file and we need to do the same for render to generate the Terraform scripts.
```

iameskild commented on apps/labs/posts/nebari-typer-cli.md:
```diff
@@ -0,0 +1,162 @@
+    <i>Output of the nebari validate and render command</i>
+</p>
+
+The render command renders all of the Terraform scripts used for the actual deployment. These Terraform scripts are broken into stages and are stored in directory called `stages`.
```
```suggestion
These Terraform scripts are broken into stages and are stored in directory called `stages`.
```

iameskild commented on apps/labs/posts/nebari-typer-cli.md:
```diff
@@ -0,0 +1,162 @@
+
+**_NOTE:_** It's not necessary to run the `validate` and `render` commands separately as both commands are run under the hood by the deploy command.
+
+Now we will see the output of the `nebari deploy` help command. The deploy command creates the provisioned resources that Nebari requires and deploys your project. Nebari also has a destroy command that works the same way the deploy works but instead of creating the provisioned resources it destroys it. The destroy command now requires confirmation, an important new feature. Previously it didn't  ...
```
```suggestion
Now we will see the output of the `nebari deploy` help command. The deploy command creates the necessary resources Nebari needs. Nebari also has a destroy command that works in a similar manner but instead of creating resources, it destroys them. The destroy command now requires confirmation, an important new feature. Previously it didn't ask for the confirmation and the cluster would start being destroyed right after executing it.
```

iameskild commented on apps/labs/posts/nebari-typer-cli.md:
```diff
@@ -0,0 +1,162 @@
+
+## 😇 My Learnings and Acknowledgements
+
+I learned a lot during my internship. First, I learned about `Typer` and `Rich` and its functionalities. Interacting with the command line interface was easy but understanding how it worked and making it more structured and user-friendly was tough intially, but playing around with the respective functions and features made it more accessible.
```
```suggestion
I learned a lot during my internship. First, I learned about `Typer` and `Rich`. Interacting with the command line interface was easy but understanding how it worked and making it more streamlined and user-friendly was challenging at first, but playing around with the respective functions and features made it more accessible.
```

iameskild commented on apps/labs/posts/nebari-typer-cli.md:
```diff
@@ -0,0 +1,162 @@
+## 😇 My Learnings and Acknowledgements
+
+I learned a lot during my internship. First, I learned about `Typer` and `Rich` and its functionalities. Interacting with the command line interface was easy but understanding how it worked and making it more structured and user-friendly was tough intially, but playing around with the respective functions and features made it more accessible.
+Nebari is a very useful platform for data scientists and reseachers to develop their projects. I hope to continue working on this very interesting project that makes life so much easier without requiring in-depth DevOps knowledge.
```
```suggestion
Nebari is a very useful platform for data scientists and researchers to perform their job duties while not requiring a dedicated DevOps administrator. I look forward to making future contributions and seeing it improve.
```

iameskild reviewed (commented):
@trallard this looks good from my perspective. Let me know if you want me to make any additional changes :)

noatamir commented:
@asmijafar20 I just noticed you didn't add any labels in your meta-section, and that your description is a tad short. For the label please use:
`category: [Developer workflows]`

For the description, I think this line from your blog is a nice summary:
`description: "The Nebari CLI consists of various commands the user needs to run to initialize, deploy, configure, and update Nebari."` This is just a suggestion, you are welcome to add to it as you see fit. 

As for your title, since we have another post with a VERY similar name, would you be open to tweaking it a bit maybe? How about "Implementing a Typer CLI for Nebari", or "Developing a Typer CLI for Nebari"? Perhaps you have another idea?

Finally, let's set the publishing date to tomorrow? If we can get all of these changes done and reviewed, I hope we can merge it then.

noatamir commented on apps/labs/posts/nebari-typer-cli.md:
```diff
@@ -0,0 +1,162 @@
+---
+title: 'Developing a Typer CLI for Nebari'
+published: October 4, 2022
```
```suggestion
published: November 7, 2022
```

asmijafar20 commented:
> @asmijafar20 I just noticed you didn't add any labels in your meta-section, and that your description is a tad short. For the label please use: `category: [Developer workflows]`
> 
> For the description, I think this line from your blog is a nice summary: `description: "The Nebari CLI consists of various commands the user needs to run to initialize, deploy, configure, and update Nebari."` This is just a suggestion, you are welcome to add to it as you see fit.
> 
> As for your title, since we have another post with a VERY similar name, would you be open to tweaking it a bit maybe? How about "Implementing a Typer CLI for Nebari", or "Developing a Typer CLI for Nebari"? Perhaps you have another idea?
> 
> Finally, let's set the publishing date to tomorrow? If we can get all of these changes done and reviewed, I hope we can merge it then.

I have added all the changes to the blog. I have also added all the images in the `svg` format because I found it more clear. I also removed video link content, we can add that once @pavithraes is done with updating my video link. I think this blog is ready to publish.

noatamir commented:
Thank you, @asmijafar20! 🌷 
Could you please remove any images you are not using from the PR? We don't need to add them to the repository if you aren't using them in your blog.

CC @trallard @iameskild I believe this is ready for your approval and publishing.

asmijafar20 commented:
> Thank you, @asmijafar20! tulip Could you please remove any images you are not using from the PR? We don't need to add them to the repository if you aren't using them in your blog.
> 
> CC @trallard @iameskild I believe this is ready for your approval and publishing.

Thanks @noatamir :) I have removed extra images.

iameskild commented:
@noatamir @asmijafar20 I just noticed that the `nebari-command` image is the older one. I just updated it with the newer version that @pavithraes created using LucidCharts :)
